baremetal/metal3machine_manager.go

```diff
@@ -1338,15 +1335,58 @@ func (m *MachineManager) SetNodeProviderID(ctx context.Context, bmhID, providerI
 		m.Log.Info("Target node is not found, requeuing")
 		return &RequeueAfterError{RequeueAfter: requeueAfter}
 	}
+	namespace := m.Metal3Machine.GetNamespace()
+	m3mName := m.Metal3Machine.GetName()
+	bmhName, err := m.getBmhNameFromM3Machine()
+	if err != nil {
+		m.Log.Info("unable to retrieve BMH from Metal3Cluster")
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	providerIDLegacy := *providerID
+	providerIDNew := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	if !m.Metal3Cluster.Spec.NoCloudProvider {
+		// Settig providerID on nodes takes time. Continuous monitoring is required.
+		nodesWithProviderIDset := 0
+		for _, node := range nodes.Items {
+			m.Log.Info("PP0: ", node.Spec.ProviderID)
+			if node.Spec.ProviderID == "" {
+				m.Log.Info("PP1: NoCloudProvider set to false. But, no providerID is set on the node either", node.GetName())
+				nodesWithProviderIDset++
```

Comment:
no need to count, since you use exit in the other cases. if you reach line 1366, it means you've gone through all nodes. If you still want to count the nodes, then you should also count in an else condition

Reply:
That was my assumption as well. However, even though all nodes are retrieved, all nodes do not have their providerID set. 
So, the reason I am counting is to make sure that the decision is based on all nodes(not just the nodes that have providerID)

Reply:
it does not matter, whether they have the providerID set or not, as long as you don't return during the loop, it means that you didn't find what you were looking for